src/overseer/store/jdbc.clj

```diff
@@ -0,0 +1,238 @@
+    ; => \"age = 30 AND email = 'bob@example.com'\""
+  [where-map]
+  (let [format-value
+        (fn [v] (if (string? v) (format "'%s'" v) v))]
+    (->> where-map
+         (map (fn [[k v]] (str (name k) " = " (format-value v))))
+         (string/join " AND "))))
+
+(defn job->jdbc-map [job]
+  {:pre [job]}
+  (let [{:keys [job/id job/type job/args job/status
+                job/failure job/heartbeat]} job]
+    (-> {:id id
+         :type (name type)
+         :status (name status)}
+        (std/when-assoc :args (some-> args std/to-edn))
+        (std/when-assoc :failure (some-> failure std/to-edn))
+        (std/when-assoc :heartbeat heartbeat))))
+
+(defn jdbc-map->job [jdbc-map]
+  {:pre [jdbc-map]}
+  (-> (std/map-kv #(keyword "job" (name %)) identity jdbc-map)
+      (util/when-update :job/status keyword)
+      (util/when-update :job/type keyword)
+      (util/when-update :job/args std/from-edn)
+      (util/when-update :job/failure std/from-edn)))
+
+(defn query-job
+  "Return a Job map for the given job-id; throws if not found"
+  [db-spec job-id]
+  (-> (j/query db-spec ["SELECT * FROM overseer_jobs WHERE id = ?" job-id])
+      first
+      (dissoc :lock_version)
+      jdbc-map->job))
+
+(defn query-lock-version [db-spec job-id]
+  (->> (j/query db-spec ["SELECT lock_version FROM overseer_jobs WHERE id = ?" job-id])
+       first
+       :lock_version))
+
+(defn update-job
+  "Update attributes of a single job using optimistic locking; raises an exception
+  if the job is stale
+
+  where-map - map of {keyword-column-name value} to scope update. id/lock_version
+              are automatically added to the final where clause
+  set-map - map of {keyword-column-name value} of new attributes to set"
+  [db-spec job-id where-map set-map]
+  (let [lock-version (query-lock-version db-spec job-id)
+        set-map' (assoc set-map :lock_version (inc lock-version))
+        where-clause [(format "id = ? AND lock_version = ? AND %s" (where-map->clause where-map)) job-id lock-version]]
+    (let [num-rows-updated (first (j/update! db-spec :overseer_jobs set-map' where-clause))]
+      (if (not= 1 num-rows-updated)
+        (throw (ex-info (str "Attempted to update stale job " (pr-str job-id)) {:type :stale-object}))
+        num-rows-updated))))
+
+(defn- direct-dependents
+  "Find the set of all Job IDs that directly depend on given `job-id`"
+  [db-spec job-id]
+  (->> (j/query db-spec ["SELECT job_id FROM overseer_dependencies WHERE dep_id = ?" job-id])
+       (map :job_id)
+       set))
+
+(defn dependents [db-spec job-id]
+  "Returns the set of all Job IDs that depend upon given `job-id`, both
+  directly and transitively"
+  (loop [all-dependents #{}
+         visited #{}
+         to-visit #{job-id}]
+    (if (empty? to-visit)
+      (set (map #(:job/id (query-job db-spec %)) all-dependents))
+      (let [job-id' (first to-visit)
+            dependents (direct-dependents db-spec job-id')
```

Comment:
Done - the graph is now traversed one level at a time, rather than individual nodes